airbyte-integrations/connectors/source-exchange-rate-api/source_exchange_rate_api/source.py

```diff
@@ -0,0 +1,154 @@
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        return None
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        yield response.json()
+
+
+class LatestRates(ExchangeRateApiStream):
+    primary_key = "timestamp"
+
+    def path(
+        self, stream_state: Mapping[str, Any] = None, stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> str:
+        return "latest"
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        params = {"base": self.config["base"]}
+        if "symbols" in self.config and self.config["symbols"]:
+            params["symbols"] = self.config["symbols"]
+        return params
+
+
+class IncrementalExchangeRateApiStream(HttpStream, IncrementalMixin, ABC):
+    state_checkpoint_interval = None
+
+    url_base = "https://api.apilayer.com/exchangerates_data/"
+
+    def __init__(self, config: Mapping[str, Any], *args, **kwargs):
+        super(IncrementalExchangeRateApiStream, self).__init__(*args, **kwargs)
+        self.config = config
+        self._cursor_value = None
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        yield response.json()
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        return None
+
+
+class HistoricalRates(IncrementalExchangeRateApiStream):
+    _date_format = "%Y-%m-%d"
+
+    cursor_field = "date"
+    primary_key = "date"
+
+    def _to_cursor_normalized_value(self, cursor_value: str) -> datetime.date:
+        return datetime.datetime.strptime(cursor_value, self._date_format).date()
+
+    def path(self, stream_state: Mapping[str, Any] = None,
+             stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None) -> str:
+        return stream_slice[self.cursor_field] if stream_slice else ""
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        params = {"base": self.config["base"]}
+        if "symbols" in self.config and self.config["symbols"]:
+            params["symbols"] = self.config["symbols"]
+        return params
+
+    @property
+    def state(self) -> Mapping[str, Any]:
+        return {self.cursor_field: self._cursor_value or self._to_cursor_normalized_value(self.config["start_date"])}
+
+    @state.setter
+    def state(self, value: Mapping[str, Any]) -> None:
+        self._cursor_value = self._to_cursor_normalized_value(value[self.cursor_field]) + datetime.timedelta(days=1)
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
```

Comment:
if I'm not mistaken, this method is an alternative to `state` getter and setter. I guess we can omit it

Reply:
Right. This method is used as a fallback if the state property is not defined or returns an error for some reason. Should I delete it?

Reply:
well, I'd prefer removing it